libraries/wallet/wallet.cpp

```diff
@@ -2244,6 +2269,84 @@ class wallet_api_impl
+      msg.meta.block = 0;
+      msg.meta.time = 0;
+      msg.signature = get_private_key( from_account.options.memo_key ).sign_compact( msg.digest() );
+      return msg;
+   }
+
+   bool verify_message( const string& message, const string& account, int block, int time, const compact_signature& sig )
+   {
+      const account_object from_account = get_account( account );
+
+      signed_message msg;
+      msg.message = message;
+      msg.meta.account = from_account.name;
+      msg.meta.memo_key = from_account.options.memo_key;
+      msg.meta.block = block;
+      msg.meta.time = time;
+      msg.signature = sig;
+
+      return verify_signed_message( msg );
+   }
+
+   bool verify_signed_message( const signed_message& message )
+   {
+      if( !message.signature.valid() ) return false;
+
+      const account_object from_account = get_account( message.meta.account );
+
+      const public_key signer( *message.signature, message.digest() );
+      if( !( message.meta.memo_key == signer ) ) return false;
+      FC_ASSERT( from_account.options.memo_key == signer,
+                 "Message was signed by contained key, but it doesn't belong to the contained account!" );
+
+      return true;
+   }
+
+   /** Verify a message signed with sign_message, in its encapsulated form.
+    *
+    * @param message the complete encapsulated message string including separators and line feeds
+    * @return true if signature matches
+    */
+   bool verify_encapsulated_message( const string& message )
+   {
+      signed_message msg;
+      size_t begin_p = message.find( ENC_HEADER );
+      FC_ASSERT( begin_p != string::npos, "BEGIN MESSAGE line not found!" );
+      size_t meta_p = message.find( ENC_META, begin_p );
+      FC_ASSERT( meta_p != string::npos, "BEGIN META line not found!" );
+      FC_ASSERT( meta_p >= begin_p + ENC_HEADER.size() + 1, "Missing message!?" );
+      size_t sig_p = message.find( ENC_SIG, meta_p );
+      FC_ASSERT( sig_p != string::npos, "BEGIN SIGNATURE line not found!" );
+      FC_ASSERT( sig_p >= meta_p + ENC_META.size(), "Missing metadata?!" );
+      size_t end_p = message.find( ENC_FOOTER, meta_p );
+      FC_ASSERT( end_p != string::npos, "END MESSAGE line not found!" );
+      FC_ASSERT( end_p >= sig_p + ENC_SIG.size() + 1, "Missing signature?!" );
+
+      msg.message = message.substr( begin_p + ENC_HEADER.size(), meta_p - begin_p - ENC_HEADER.size() - 1 );
+      const string meta = message.substr( meta_p + ENC_META.size(), sig_p - meta_p - ENC_META.size() );
+      const string sig = message.substr( sig_p + ENC_SIG.size(), end_p - sig_p - ENC_SIG.size() - 1 );
+
+      msg.meta.account = meta_extract( meta, "account" );
+      msg.meta.memo_key = public_key_type( meta_extract( meta, "memokey" ) );
+      msg.meta.block = boost::lexical_cast<uint32_t>( meta_extract( meta, "block" ) );
+      msg.meta.time = boost::lexical_cast<uint64_t>( meta_extract( meta, "timestamp" ) );
```

Comment:
This can fail, because specification (precision and format) of timestamp is not well-defined.

Reply:
Switched "time" field to string.